File: loans-service/Audit/AuditService.cs
using Elastic.Clients.Elasticsearch;
using Elastic.Clients.Elasticsearch.QueryDsl;
using loans_service.Middleware;
using System.Security.Claims;

namespace loans_service.Audit;

public class AuditService
{
    public const string IndexName = "loan-audit";

    private readonly ElasticsearchClient _client;
    private readonly IHttpContextAccessor _httpContextAccessor;
    private readonly ILogger<AuditService> _logger;

    public AuditService(
        ElasticsearchClient client,
        IHttpContextAccessor httpContextAccessor,
        ILogger<AuditService> logger)
    {
        _client = client;
        _httpContextAccessor = httpContextAccessor;
        _logger = logger;
    }

    public async Task RecordAsync(
        string eventType,
        string aggregate,
        string aggregateId,
        object? before,
        object? after,
        CancellationToken cancellationToken)
    {
        var entry = new AuditEntry
        {
            EventType = eventType,
            Aggregate = aggregate,
            AggregateId = aggregateId,
            ActorId = _httpContextAccessor.HttpContext?.User.FindFirst("sub")?.Value,
            CorrelationId = CorrelationContext.CorrelationId,
            OccurredAt = DateTime.UtcNow,
            Before = before,
            After = after,
        };

        var response = await _client.IndexAsync(
            entry,
            i => i.Index(IndexName).Id(Guid.NewGuid().ToString()),
            cancellationToken);

        if (!response.IsSuccess())
        {
            throw new InvalidOperationException(
                $"Failed to record audit entry for {aggregate} {aggregateId}: {response.DebugInformation}");
        }

        _logger.LogInformation(
            "Recorded {EventType} for {Aggregate} {AggregateId} (correlation {CorrelationId})",
            eventType, aggregate, aggregateId, entry.CorrelationId);
    }

    public async Task<(IReadOnlyList<AuditEntry> Hits, long Total)> SearchAsync(
        string? aggregateId,
        string? eventType,
        DateTime? from,
        DateTime? to,
        int page,
        int pageSize,
        CancellationToken cancellationToken)
    {
        var must = new List<Query>();

        if (!string.IsNullOrWhiteSpace(aggregateId))
        {
            must.Add(new TermQuery { Field = "aggregateId", Value = aggregateId });
        }

        if (!string.IsNullOrWhiteSpace(eventType))
        {
            must.Add(new TermQuery { Field = "eventType", Value = eventType });
        }

        if (from.HasValue || to.HasValue)
        {
            var range = new DateRangeQuery("occurredAt");
            if (from.HasValue) range.Gte = from.Value;
            if (to.HasValue) range.Lte = to.Value;
            must.Add(range);
        }

        Query query = must.Count > 0 ? new BoolQuery { Must = must } : new MatchAllQuery();

        var response = await _client.SearchAsync<AuditEntry>(s => s
            .Indices(IndexName)
            .Query(query)
            .Sort(sort => sort
                .Field("occurredAt", f => f.Order(SortOrder.Desc)))
            .From((page - 1) * pageSize)
            .Size(pageSize),
            cancellationToken);

        if (!response.IsSuccess())
        {
            throw new InvalidOperationException(
                $"Failed to search audit trail: {response.DebugInformation}");
        }

        return (response.Hits.Select(h => h.Source).Where(h => h is not null).Cast<AuditEntry>().ToList(), response.Total);
    }

    public async Task EnsureIndexAsync(CancellationToken cancellationToken)
    {
        var exists = await _client.Indices.ExistsAsync(IndexName, cancellationToken);

        if (exists.Exists)
        {
            return;
        }

        var response = await _client.Indices.CreateAsync<AuditEntry>(IndexName,
            c => c.Mappings(m => m
                .Properties(p => p
                    .Keyword(k => k.EventType)
                    .Keyword(k => k.Aggregate)
                    .Keyword(k => k.AggregateId)
                    .Keyword(k => k.ActorId)
                    .Keyword(k => k.CorrelationId)
                    .Date(d => d.OccurredAt))),
            cancellationToken);

        if (!response.IsSuccess())
        {
            throw new InvalidOperationException(
                $"Failed to create audit index {IndexName}: {response.DebugInformation}");
        }

        _logger.LogInformation("Created Elasticsearch index {Index}", IndexName);
    }
}

File: accounts-service/src/common/audit/audit.service.ts
import {
  Injectable,
  Logger,
  OnModuleDestroy,
  OnModuleInit,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Client } from '@elastic/elasticsearch';

export interface AuditEntry {
  eventType: 'account.created' | 'account.updated' | 'account.deleted';
  aggregate: 'account';
  aggregateId: string;
  actorId?: string | null;
  correlationId?: string | null;
  occurredAt: Date;
  before?: Record<string, unknown> | null;
  after?: Record<string, unknown> | null;
}

export interface AuditSearchParams {
  aggregateId?: string;
  eventType?: string;
  from?: Date;
  to?: Date;
  page?: number;
  pageSize?: number;
}

@Injectable()
export class AuditService implements OnModuleInit, OnModuleDestroy {
  public static readonly IndexName = 'account-audit';

  private readonly logger = new Logger(AuditService.name);
  private readonly client: Client;

  constructor(configService: ConfigService) {
    const node =
      configService.get<string>('ELASTICSEARCH_URL') ?? 'http://localhost:9200';
    this.client = new Client({ node });
  }

  async onModuleInit(): Promise<void> {
    await this.ensureIndex();
  }

  async onModuleDestroy(): Promise<void> {
    await this.client.close();
  }

  async record(entry: AuditEntry): Promise<void> {
    const document = {
      ...entry,
      occurredAt: entry.occurredAt.toISOString(),
    };

    await this.client.index({
      index: AuditService.IndexName,
      id: crypto.randomUUID(),
      document,
      refresh: 'wait_for',
    });

    this.logger.log(
      `Recorded ${entry.eventType} for ${entry.aggregate} ${entry.aggregateId} (correlation ${entry.correlationId ?? 'n/a'})`,
    );
  }

  async search(
    params: AuditSearchParams,
  ): Promise<{ hits: AuditEntry[]; total: number }> {
    const page = Math.max(params.page ?? 1, 1);
    const pageSize = Math.min(Math.max(params.pageSize ?? 20, 1), 100);

    const must: Array<Record<string, unknown>> = [];

    if (params.aggregateId) {
      must.push({ term: { aggregateId: params.aggregateId } });
    }
    if (params.eventType) {
      must.push({ term: { eventType: params.eventType } });
    }
    if (params.from || params.to) {
      const range: Record<string, unknown> = {};
      if (params.from) range.gte = params.from.toISOString();
      if (params.to) range.lte = params.to.toISOString();
      must.push({ range: { occurredAt: range } });
    }

    const query = must.length > 0 ? { bool: { must } } : { match_all: {} };

    const response = await this.client.search({
      index: AuditService.IndexName,
      query,
      sort: [{ occurredAt: { order: 'desc' } }],
      from: (page - 1) * pageSize,
      size: pageSize,
    });

    const hits = response.hits.hits.map((hit) => hit._source as AuditEntry);
    const total =
      typeof response.hits.total === 'number'
        ? response.hits.total
        : (response.hits.total?.value ?? 0);

    return { hits, total };
  }

  private async ensureIndex(): Promise<void> {
    const exists = await this.client.indices.exists({
      index: AuditService.IndexName,
    });

    if (exists) {
      return;
    }

    await this.client.indices.create({
      index: AuditService.IndexName,
      mappings: {
        properties: {
          eventType: { type: 'keyword' },
          aggregate: { type: 'keyword' },
          aggregateId: { type: 'keyword' },
          actorId: { type: 'keyword' },
          correlationId: { type: 'keyword' },
          occurredAt: { type: 'date' },
          before: { type: 'object', enabled: false },
          after: { type: 'object', enabled: false },
        },
      },
    });

    this.logger.log(`Created Elasticsearch index ${AuditService.IndexName}`);
  }
}
